src/v/kafka/requests/fetch_request.cc

```diff
@@ -492,18 +515,35 @@ static ss::future<> fetch_topic_partition(
  * response must be reassembled such that the responses appear in these
  * order as the partitions in the request.
  */
+
 static ss::future<> fetch_topic_partitions(op_context& octx) {
-    return ss::do_for_each(
-             octx.request.cbegin(),
-             octx.request.cend(),
-             [&octx](const fetch_request::const_iterator::value_type& p) {
-                 return fetch_topic_partition(octx, p).then([&octx] {
-                     if (!octx.initial_fetch) {
-                         octx.response_iterator++;
-                     }
-                 });
-             })
-      .then([&octx] { octx.reset_context(); });
+    static constexpr std::chrono::milliseconds fetch_reads_debounce_timeout
+      = std::chrono::milliseconds(5);
+
+    auto resp_it = octx.response_begin();
+    std::vector<ss::future<>> fetches;
+    std::transform(
+      octx.request.cbegin(),
+      octx.request.cend(),
+      std::back_inserter(fetches),
+      [&resp_it, &octx](const fetch_request::const_iterator::value_type& p) {
+          // we use gate as we may not wait for all the fetch results
+          return fetch_topic_partition(octx, p, resp_it++);
+      });
+
+    return ss::do_with(
+      std::move(fetches), [&octx](std::vector<ss::future<>>& fetches) {
+          return ss::when_all_succeed(fetches.begin(), fetches.end())
+            .then([&octx] {
+                if (octx.should_stop_fetch()) {
+                    return ss::now();
+                }
+                octx.reset_context();
+                // debounce next read retry
+                return ss::sleep(std::min(
```

Comment:
use `sleep_abortable()` and pass in the context `abort_source`

Reply:
we do not have `abort_source` in `request_context`. This is very short sleep, do we need to abort it ?

Reply:
let's defer, please create a ticket.